sensenet/models/wrappers.py

```diff
@@ -193,17 +186,102 @@ def bigml_resource(resource):
+def model_from_dictionary(model_dict, settings):
     settings_object = ensure_settings(settings)
 
-    if bigml_resource(model):
-        return create_model(bigml_resource(model), settings=settings)
-    elif is_deepnet(model):
+    if bigml_resource(model_dict):
+        model = bigml_resource(model)
+    else:
+        model = model_dict
+
+    if is_deepnet(model):
         if is_yolo_model(model):
             return ObjectDetector(model, settings_object)
         else:
             return Deepnet(model, settings_object)
     elif isinstance(model, dict):
         raise ValueError('Model format not recognized: %s' % str(model.keys()))
+
+def model_from_bundle(bundle_file):
+    bundle_name = os.path.basename(bundle_file)
+
+    with tempfile.TemporaryDirectory() as saved_model_temp:
+        temp_bundle = os.path.join(saved_model_temp, bundle_name)
+        shutil.copyfile(bundle_file, temp_bundle)
+        model_dir = read_bundle(temp_bundle)
+
+        model = tf.keras.models.load_model(model_dir)
+        settings_path = os.path.join(model_dir, SETTINGS_PATH)
+
+        if os.path.exists(settings_path):
+            with open(settings_path, 'r') as fin:
+                settings = json.load(fin)
+        else:
+            settings = None
+
+    dtype = settings['deepnet_type']
+
+    if dtype == 'ObjectDetector':
+        return ObjectDetector(model, settings)
+    elif dtype == 'Deepnet':
+        return Deepnet(model, settings)
+    else:
+        raise ValueError('Invalid deepnet type: "%s"' % dtype)
+
+def create_model(anobject, settings=None):
+    if isinstance(anobject, str):
+        if os.path.exists(anobject):
+            if anobject.endswith(BUNDLE_EXTENSION):
+                return model_from_bundle(anobject)
+            else:
+                with open(anobject, 'r') as fin:
+                    return model_from_dictionary(json.load(fin), settings)
+        else:
+            raise IOError('File %s not found' % str(anobject))
+    elif isinstance(anobject, dict):
+        return model_from_dictionary(anobject, settings)
+    else:
+        raise TypeError('Input argument cannot be a %s' % str(type(anobject)))
+
+def convert(model, settings, output_path, to_format):
+    """Convert some structure describing a wrapped model to a given output
+    format.
+
+    The first input to this function can be one of the model wrapper
+    classes (e.g., ObjectDetector), a dict used to instantiate a
+    wrapper, the path to a file containing JSON for such a dict, or
+    the path to a bundled saved model.
+
+    The second a dict of settings for the model, which can contain
+    things like the IOU threshold for non-max suppression (see
+    sensenet.model.settings).  Note that if the first argument is an
+    already-instantiated model, this argument is ignored.
+
+    The third is the path to which to output the converted model.
+
+    The fourth is the format to which to convert the model, any of
```

Comment:
smbundle files are the new name for JSON files containing Pyramid/Sensenet models, am I wrong?
Will be the functions that read smbundles able to read models directly downloaded from apian?

Reply:
> smbundle files are the new name for JSON files containing Pyramid/Sensenet models, am I wrong?

No, the JSON format is still the primary format that will get passed around to and from clients.  This smbundle thing is basically just a byte-wise concatenation of all of the files in a TF SavedModel directory so we can run the model on the Java bindings without having to parse the JSON and instantiate the whole thing layer by layer in java for wintermute predictions (we can just read the SavedModel using the TF Java bindings).

It is much lighter weight than the raw JSON, so you could pass around this format in your projects if you wanted to.  Again, it's just a thin wrapper for SavedModel.